File: sidecars/minecraft/src/protocol.ts
import { z } from 'zod';

export const PROTOCOL_NAME = 'gamma.minecraft' as const;
export const PROTOCOL_VERSION = 1 as const;
export const MAX_COMMAND_DEADLINE_SECONDS = 900 as const;

export const MESSAGE_TYPES = [
  'hello',
  'welcome',
  'heartbeat',
  'sidecar_status',
  'minecraft_status',
  'command',
  'command_ack',
  'command_progress',
  'state_snapshot',
  'terminal_result',
  'cancel_command',
  'emergency_stop',
  'shutdown',
  'protocol_error'
] as const;

export const COMMAND_NAMES = [
  'join',
  'leave',
  'follow_owner',
  'wait_here',
  'come_here',
  'look_at_owner',
  'report_status',
  'stop',
  'emergency_stop'
] as const;

export const TERMINAL_OUTCOMES = [
  'completed',
  'cancelled',
  'failed',
  'rejected',
  'timed_out'
] as const;

export const FAILURE_CODES = [
  'FEATURE_DISABLED',
  'SIDECAR_UNAVAILABLE',
  'PROTOCOL_MISMATCH',
  'MINECRAFT_NOT_CONNECTED',
  'OWNER_NOT_CONFIGURED',
  'OWNER_NOT_PRESENT',
  'UNAUTHORIZED_REQUESTER',
  'INVALID_COMMAND',
  'INVALID_STATE',
  'COMMAND_ALREADY_ACTIVE',
  'DESTINATION_UNAVAILABLE',
  'PATH_NOT_FOUND',
  'PATH_STALLED',
  'OWNER_TOO_FAR_AWAY',
  'DEADLINE_EXCEEDED',
  'SAFETY_POLICY_BLOCKED',
  'UNSUPPORTED_DIMENSION',
  'SIDECAR_DISCONNECTED',
  'MINECRAFT_SERVER_DISCONNECTED',
  'BOT_DEAD',
  'EMERGENCY_STOP_ACTIVE',
  'INTERNAL_SIDECAR_ERROR',
  'INTERNAL_GAMMA_ERROR'
] as const;

export const COMPANION_STATES = [
  'DISCONNECTED',
  'IDLE',
  'FOLLOWING',
  'WAITING',
  'RETURNING',
  'FLEEING',
  'DEAD',
  'STOPPED'
] as const;

export type MessageType = (typeof MESSAGE_TYPES)[number];
export type CommandName = (typeof COMMAND_NAMES)[number];
export type TerminalOutcome = (typeof TERMINAL_OUTCOMES)[number];
export type FailureCode = (typeof FAILURE_CODES)[number];
export type CompanionState = (typeof COMPANION_STATES)[number];

const IDENTIFIER_PATTERN = /^[A-Za-z0-9][A-Za-z0-9._:-]*$/;
const PROFILE_ID_PATTERN = /^[a-z0-9][a-z0-9._-]*$/;
const SAFE_TEXT_PATTERN = /^[^\x00-\x1f\x7f]+$/;
const VERSION_PATTERN = /^[A-Za-z0-9][A-Za-z0-9.+_-]*$/;
const UTC_TIMESTAMP_PATTERN =
  /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.\d+)?(?:Z|[+-]00:00)$/;

const boundedIdSchema = z.string().min(1).max(128).regex(IDENTIFIER_PATTERN);
const profileIdSchema = z.string().min(1).max(64).regex(PROFILE_ID_PATTERN);
const safeReasonSchema = z.string().trim().min(1).max(160).regex(SAFE_TEXT_PATTERN);
const safeDetailSchema = z.string().trim().min(1).max(256).regex(SAFE_TEXT_PATTERN);
const safeDisplayNameSchema = z.string().trim().min(1).max(64).regex(SAFE_TEXT_PATTERN);
const boundedVersionSchema = z.string().min(1).max(32).regex(VERSION_PATTERN);

function isUtcTimestamp(value: string): boolean {
  const match = UTC_TIMESTAMP_PATTERN.exec(value);
  if (match === null) return false;
  const parts = match.slice(1, 7).map(Number);
  const [year, month, day, hour, minute, second] = parts;
  if (
    year === undefined ||
    month === undefined ||
    day === undefined ||
    hour === undefined ||
    minute === undefined ||
    second === undefined ||
    year < 1 ||
    month < 1 ||
    month > 12 ||
    day < 1 ||
    hour > 23 ||
    minute > 59 ||
    second > 59
  ) {
    return false;
  }
  const parsed = new Date(0);
  parsed.setUTCFullYear(year, month - 1, day);
  parsed.setUTCHours(hour, minute, second, 0);
  return (
    parsed.getUTCFullYear() === year &&
    parsed.getUTCMonth() === month - 1 &&
    parsed.getUTCDate() === day &&
    parsed.getUTCHours() === hour &&
    parsed.getUTCMinutes() === minute &&
    parsed.getUTCSeconds() === second &&
    Number.isFinite(Date.parse(value))
  );
}

const utcTimestampSchema = z.string().refine(isUtcTimestamp, 'Expected a valid UTC timestamp');
const sequenceSchema = z.number().int().min(0);
const commandNameSchema = z.enum(COMMAND_NAMES);
const terminalOutcomeSchema = z.enum(TERMINAL_OUTCOMES);
const failureCodeSchema = z.enum(FAILURE_CODES);
const companionStateSchema = z.enum(COMPANION_STATES);
const sidecarConnectionStateSchema = z.enum(['connecting', 'connected', 'stale', 'disconnected']);
const minecraftConnectionStateSchema = z.enum(['disconnected', 'connecting', 'connected']);
const minecraftDimensionSchema = z.enum([
  'minecraft:overworld',
  'minecraft:the_nether',
  'minecraft:the_end'
]);

const failureSchema = z.strictObject({
  code: failureCodeSchema,
  safe_detail: safeDetailSchema.nullish(),
  retriable: z.boolean()
});

const roundedPositionSchema = z.strictObject({
  x: z.number().int().min(-30_000_000).max(30_000_000),
  y: z.number().int().min(-2_048).max(2_048),
  z: z.number().int().min(-30_000_000).max(30_000_000)
});

const joinArgumentsSchema = z.strictObject({
  connection_profile_id: profileIdSchema.nullish()
});
const leaveArgumentsSchema = z.strictObject({ reason: safeReasonSchema.nullish() });
const followOwnerArgumentsSchema = z.strictObject({
  follow_distance: z.number().min(2).max(6).default(3),
  lease_duration_seconds: z.number().int().min(5).max(900).default(300)
});
const waitHereArgumentsSchema = z.strictObject({ reason: safeReasonSchema.nullish() });
const comeHereArgumentsSchema = z.strictObject({
  arrival_distance: z.number().min(2).max(4).default(3)
});
const lookAtOwnerArgumentsSchema = z.strictObject({
  duration_seconds: z.number().min(0.1).max(10).default(2)
});
const reportStatusArgumentsSchema = z.strictObject({
  detail_level: z.enum(['basic', 'standard']).default('standard')
});
const stopArgumentsSchema = z.strictObject({ reason: safeReasonSchema.nullish() });
const emergencyStopArgumentsSchema = z.strictObject({ reason: safeReasonSchema.nullish() });

const commandPayloadSchema = z.discriminatedUnion('name', [
  z.strictObject({
    name: z.literal('join'),
    deadline_at: utcTimestampSchema,
    arguments: joinArgumentsSchema.default({})
  }),
  z.strictObject({
    name: z.literal('leave'),
    deadline_at: utcTimestampSchema,
    arguments: leaveArgumentsSchema.default({})
  }),
  z.strictObject({
    name: z.literal('follow_owner'),
    deadline_at: utcTimestampSchema,
    arguments: followOwnerArgumentsSchema.default({
      follow_distance: 3,
      lease_duration_seconds: 300
    })
  }),
  z.strictObject({
    name: z.literal('wait_here'),
    deadline_at: utcTimestampSchema,
    arguments: waitHereArgumentsSchema.default({})
  }),
  z.strictObject({
    name: z.literal('come_here'),
    deadline_at: utcTimestampSchema,
    arguments: comeHereArgumentsSchema.default({ arrival_distance: 3 })
  }),
  z.strictObject({
    name: z.literal('look_at_owner'),
    deadline_at: utcTimestampSchema,
    arguments: lookAtOwnerArgumentsSchema.default({ duration_seconds: 2 })
  }),
  z.strictObject({
    name: z.literal('report_status'),
    deadline_at: utcTimestampSchema,
    arguments: reportStatusArgumentsSchema.default({ detail_level: 'standard' })
  }),
  z.strictObject({
    name: z.literal('stop'),
    deadline_at: utcTimestampSchema,
    arguments: stopArgumentsSchema.default({})
  }),
  z.strictObject({
    name: z.literal('emergency_stop'),
    deadline_at: utcTimestampSchema,
    arguments: emergencyStopArgumentsSchema.default({})
  })
]);

const commonEnvelopeShape = {
  protocol: z.literal(PROTOCOL_NAME),
  version: z.literal(PROTOCOL_VERSION),
  message_id: boundedIdSchema,
  sent_at: utcTimestampSchema,
  sequence: sequenceSchema
};

const connectedEnvelopeShape = {
  ...commonEnvelopeShape,
  connection_id: boundedIdSchema
};

const commandEnvelopeShape = {
  ...connectedEnvelopeShape,
  trace_id: boundedIdSchema,
  command_id: boundedIdSchema
};

export const helloMessageSchema = z.strictObject({
  ...commonEnvelopeShape,
  type: z.literal('hello'),
  payload: z.strictObject({
    supported_versions: z.array(z.literal(1)).min(1).max(4),
    sidecar_instance_id: boundedIdSchema,
    sidecar_build: boundedVersionSchema,
    capabilities: z.array(z.literal('companion_v1')).min(1).max(4),
    node_version: boundedVersionSchema,
    minecraft_library_version: boundedVersionSchema,
    pathfinder_version: boundedVersionSchema,
    companion_state: z.enum(['DISCONNECTED', 'IDLE'])
  })
});

export const welcomeMessageSchema = z.strictObject({
  ...connectedEnvelopeShape,
  type: z.literal('welcome'),
  payload: z.strictObject({
    selected_version: z.literal(1),
    heartbeat_interval_seconds: z.number().int().min(1).max(60),
    liveness_timeout_seconds: z.number().int().min(2).max(180),
    maximum_message_bytes: z.number().int().min(1_024).max(1_048_576),
    command_cache_ttl_seconds: z.number().int().min(60).max(86_400),
    command_cache_capacity: z.number().int().min(1).max(10_000),
    minecraft_chat_output_enabled: z.literal(false).default(false)
  })
});

export const heartbeatMessageSchema = z.strictObject({
  ...connectedEnvelopeShape,
  type: z.literal('heartbeat'),
  payload: z.strictObject({
    companion_state: companionStateSchema,
    last_received_sequence: sequenceSchema.nullish()
  })
});

export const sidecarStatusMessageSchema = z.strictObject({
  ...connectedEnvelopeShape,
  type: z.literal('sidecar_status'),
  payload: z.strictObject({
    connection_state: sidecarConnectionStateSchema,
    companion_state: companionStateSchema,
    uptime_seconds: z.number().int().min(0).max(31_536_000),
    last_failure: failureSchema.nullish()
  })
});

export const minecraftStatusMessageSchema = z.strictObject({
  ...connectedEnvelopeShape,
  type: z.literal('minecraft_status'),
  payload: z.strictObject({
    connection_state: minecraftConnectionStateSchema,
    companion_state: companionStateSchema,
    negotiated_version: boundedVersionSchema.nullish(),
    dimension: minecraftDimensionSchema.nullish()
  })
});

export const commandMessageSchema = z.strictObject({
  ...commandEnvelopeShape,
  type: z.literal('command'),
  payload: commandPayloadSchema
});

const commandAckPayloadSchema = z
  .strictObject({
    accepted: z.boolean(),
    command_name: commandNameSchema,
    failure: failureSchema.nullish()
  })
  .superRefine((payload, context) => {
    if (payload.accepted && payload.failure != null) {
      context.addIssue({ code: 'custom', message: 'Accepted acknowledgments cannot contain a failure' });
    }
    if (!payload.accepted && payload.failure == null) {
      context.addIssue({ code: 'custom', message: 'Rejected acknowledgments require a failure' });
    }
  });

export const commandAckMessageSchema = z.strictObject({
  ...commandEnvelopeShape,
  type: z.literal('command_ack'),
  payload: commandAckPayloadSchema
});

export const commandProgressMessageSchema = z.strictObject({
  ...commandEnvelopeShape,
  type: z.literal('command_progress'),
  payload: z.strictObject({
    command_name: commandNameSchema,
    phase: z.enum(['started', 'moving', 'waiting', 'retrying']),
    elapsed_ms: z.number().int().min(0).max(86_400_000),
    safe_detail: safeDetailSchema.nullish()
  })
});

const stateSnapshotPayloadSchema = z
  .strictObject({
    sidecar_connection_state: sidecarConnectionStateSchema,
    minecraft_connection_state: minecraftConnectionStateSchema,
    companion_state: companionStateSchema,
    owner_present: z.boolean(),
    owner_display_name: safeDisplayNameSchema.nullish(),
    owner_uuid: z.uuid().nullish(),
    dimension: minecraftDimensionSchema.nullish(),
    rounded_position: roundedPositionSchema.nullish(),
    health: z.number().min(0).max(20).nullish(),
    hunger: z.number().int().min(0).max(20).nullish(),
    active_command_id: boundedIdSchema.nullish(),
    active_command_name: commandNameSchema.nullish(),
    last_terminal_outcome: terminalOutcomeSchema.nullish(),
    last_failure_code: failureCodeSchema.nullish()
  })
  .superRefine((payload, context) => {
    if ((payload.active_command_id == null) !== (payload.active_command_name == null)) {
      context.addIssue({
        code: 'custom',
        message: 'Active command identifier and name must appear together'
      });
    }
  });

export const stateSnapshotMessageSchema = z.strictObject({
  ...connectedEnvelopeShape,
  type: z.literal('state_snapshot'),
  trace_id: boundedIdSchema.nullish(),
  command_id: boundedIdSchema.nullish(),
  payload: stateSnapshotPayloadSchema
});

const terminalResultPayloadSchema = z
  .strictObject({
    command_name: commandNameSchema,
    outcome: terminalOutcomeSchema,
    elapsed_ms: z.number().int().min(0).max(86_400_000),
    safe_detail: safeDetailSchema.nullish(),
    failure: failureSchema.nullish()
  })
  .superRefine((payload, context) => {
    const requiresFailure = ['failed', 'rejected', 'timed_out'].includes(payload.outcome);
    if (requiresFailure && payload.failure == null) {
      context.addIssue({ code: 'custom', message: 'This terminal outcome requires a failure' });
    }
    if (!requiresFailure && payload.failure != null) {
      context.addIssue({ code: 'custom', message: 'This terminal outcome cannot contain a failure' });
    }
  });

export const terminalResultMessageSchema = z.strictObject({
  ...commandEnvelopeShape,
  type: z.literal('terminal_result'),
  payload: terminalResultPayloadSchema
});

export const cancelCommandMessageSchema = z.strictObject({
  ...commandEnvelopeShape,
  type: z.literal('cancel_command'),
  payload: z.strictObject({ reason: safeReasonSchema.nullish() })
});

export const emergencyStopMessageSchema = z.strictObject({
  ...commandEnvelopeShape,
  type: z.literal('emergency_stop'),
  payload: z.strictObject({ reason: safeReasonSchema.nullish() })
});

export const shutdownMessageSchema = z.strictObject({
  ...connectedEnvelopeShape,
  type: z.literal('shutdown'),
  trace_id: boundedIdSchema,
  payload: z.strictObject({
    reason: safeReasonSchema.nullish(),
    leave_minecraft: z.literal(true).default(true)
  })
});

export const protocolErrorMessageSchema = z.strictObject({
  ...commonEnvelopeShape,
  type: z.literal('protocol_error'),
  connection_id: boundedIdSchema.nullish(),
  trace_id: boundedIdSchema.nullish(),
  command_id: boundedIdSchema.nullish(),
  payload: z.strictObject({
    code: failureCodeSchema,
    safe_detail: safeDetailSchema,
    retriable: z.boolean(),
    offending_type: z.string().min(1).max(64).nullish()
  })
});

const protocolMessageUnionSchema = z.discriminatedUnion('type', [
  helloMessageSchema,
  welcomeMessageSchema,
  heartbeatMessageSchema,
  sidecarStatusMessageSchema,
  minecraftStatusMessageSchema,
  commandMessageSchema,
  commandAckMessageSchema,
  commandProgressMessageSchema,
  stateSnapshotMessageSchema,
  terminalResultMessageSchema,
  cancelCommandMessageSchema,
  emergencyStopMessageSchema,
  shutdownMessageSchema,
  protocolErrorMessageSchema
]);

export const protocolMessageSchema = protocolMessageUnionSchema.superRefine((message, context) => {
  if (message.type === 'command') {
    const sentAt = Date.parse(message.sent_at);
    const deadlineAt = Date.parse(message.payload.deadline_at);
    const delta = deadlineAt - sentAt;
    if (delta <= 0) {
      context.addIssue({ code: 'custom', message: 'deadline_at must be later than sent_at' });
    } else if (delta > MAX_COMMAND_DEADLINE_SECONDS * 1_000) {
      context.addIssue({
        code: 'custom',
        message: 'deadline_at cannot be more than 900 seconds after sent_at'
      });
    }
  }
  if (
    message.type === 'state_snapshot' &&
    message.command_id != null &&
    message.trace_id == null
  ) {
    context.addIssue({
      code: 'custom',
      message: 'Command-correlated snapshots require trace_id'
    });
  }
});

export type HelloMessage = z.infer<typeof helloMessageSchema>;
export type WelcomeMessage = z.infer<typeof welcomeMessageSchema>;
export type HeartbeatMessage = z.infer<typeof heartbeatMessageSchema>;
export type SidecarStatusMessage = z.infer<typeof sidecarStatusMessageSchema>;
export type MinecraftStatusMessage = z.infer<typeof minecraftStatusMessageSchema>;
export type CommandMessage = z.infer<typeof commandMessageSchema>;
export type CommandAckMessage = z.infer<typeof commandAckMessageSchema>;
export type CommandProgressMessage = z.infer<typeof commandProgressMessageSchema>;
export type StateSnapshotMessage = z.infer<typeof stateSnapshotMessageSchema>;
export type TerminalResultMessage = z.infer<typeof terminalResultMessageSchema>;
export type CancelCommandMessage = z.infer<typeof cancelCommandMessageSchema>;
export type EmergencyStopMessage = z.infer<typeof emergencyStopMessageSchema>;
export type ShutdownMessage = z.infer<typeof shutdownMessageSchema>;
export type ProtocolErrorMessage = z.infer<typeof protocolErrorMessageSchema>;
export type ProtocolMessage = z.infer<typeof protocolMessageSchema>;

export function parseProtocolMessage(input: unknown): ProtocolMessage {
  return protocolMessageSchema.parse(input);
}

File: src/gamma/integrations/minecraft/protocol.py
"""Strict, runtime-independent Minecraft companion protocol v1 models."""

from __future__ import annotations

from datetime import datetime, timedelta
from enum import Enum
from typing import Annotated, Literal, Self
from uuid import UUID

from pydantic import (
    AwareDatetime,
    BaseModel,
    ConfigDict,
    Field,
    StringConstraints,
    TypeAdapter,
    field_validator,
    model_validator,
)


PROTOCOL_NAME = "gamma.minecraft"
PROTOCOL_VERSION = 1
MAX_COMMAND_DEADLINE_SECONDS = 900

BoundedId = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=128,
        pattern=r"^[A-Za-z0-9][A-Za-z0-9._:-]*$",
    ),
]
ProfileId = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=64,
        pattern=r"^[a-z0-9][a-z0-9._-]*$",
    ),
]
SafeReason = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=160,
        strip_whitespace=True,
        pattern=r"^[^\x00-\x1f\x7f]+$",
    ),
]
SafeDetail = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=256,
        strip_whitespace=True,
        pattern=r"^[^\x00-\x1f\x7f]+$",
    ),
]
SafeDisplayName = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=64,
        strip_whitespace=True,
        pattern=r"^[^\x00-\x1f\x7f]+$",
    ),
]
BoundedVersion = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=32,
        pattern=r"^[A-Za-z0-9][A-Za-z0-9.+_-]*$",
    ),
]


class CommandName(str, Enum):
    JOIN = "join"
    LEAVE = "leave"
    FOLLOW_OWNER = "follow_owner"
    WAIT_HERE = "wait_here"
    COME_HERE = "come_here"
    LOOK_AT_OWNER = "look_at_owner"
    REPORT_STATUS = "report_status"
    STOP = "stop"
    EMERGENCY_STOP = "emergency_stop"


class TerminalOutcome(str, Enum):
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    FAILED = "failed"
    REJECTED = "rejected"
    TIMED_OUT = "timed_out"


class FailureCode(str, Enum):
    FEATURE_DISABLED = "FEATURE_DISABLED"
    SIDECAR_UNAVAILABLE = "SIDECAR_UNAVAILABLE"
    PROTOCOL_MISMATCH = "PROTOCOL_MISMATCH"
    MINECRAFT_NOT_CONNECTED = "MINECRAFT_NOT_CONNECTED"
    OWNER_NOT_CONFIGURED = "OWNER_NOT_CONFIGURED"
    OWNER_NOT_PRESENT = "OWNER_NOT_PRESENT"
    UNAUTHORIZED_REQUESTER = "UNAUTHORIZED_REQUESTER"
    INVALID_COMMAND = "INVALID_COMMAND"
    INVALID_STATE = "INVALID_STATE"
    COMMAND_ALREADY_ACTIVE = "COMMAND_ALREADY_ACTIVE"
    DESTINATION_UNAVAILABLE = "DESTINATION_UNAVAILABLE"
    PATH_NOT_FOUND = "PATH_NOT_FOUND"
    PATH_STALLED = "PATH_STALLED"
    OWNER_TOO_FAR_AWAY = "OWNER_TOO_FAR_AWAY"
    DEADLINE_EXCEEDED = "DEADLINE_EXCEEDED"
    SAFETY_POLICY_BLOCKED = "SAFETY_POLICY_BLOCKED"
    UNSUPPORTED_DIMENSION = "UNSUPPORTED_DIMENSION"
    SIDECAR_DISCONNECTED = "SIDECAR_DISCONNECTED"
    MINECRAFT_SERVER_DISCONNECTED = "MINECRAFT_SERVER_DISCONNECTED"
    BOT_DEAD = "BOT_DEAD"
    EMERGENCY_STOP_ACTIVE = "EMERGENCY_STOP_ACTIVE"
    INTERNAL_SIDECAR_ERROR = "INTERNAL_SIDECAR_ERROR"
    INTERNAL_GAMMA_ERROR = "INTERNAL_GAMMA_ERROR"


class CompanionState(str, Enum):
    DISCONNECTED = "DISCONNECTED"
    IDLE = "IDLE"
    FOLLOWING = "FOLLOWING"
    WAITING = "WAITING"
    RETURNING = "RETURNING"
    FLEEING = "FLEEING"
    DEAD = "DEAD"
    STOPPED = "STOPPED"


class SidecarConnectionState(str, Enum):
    CONNECTING = "connecting"
    CONNECTED = "connected"
    STALE = "stale"
    DISCONNECTED = "disconnected"


class MinecraftConnectionState(str, Enum):
    DISCONNECTED = "disconnected"
    CONNECTING = "connecting"
    CONNECTED = "connected"


class MinecraftDimension(str, Enum):
    OVERWORLD = "minecraft:overworld"
    NETHER = "minecraft:the_nether"
    END = "minecraft:the_end"


class SafeModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Failure(SafeModel):
    code: FailureCode
    safe_detail: SafeDetail | None = None
    retriable: bool


class RoundedPosition(SafeModel):
    x: int = Field(ge=-30_000_000, le=30_000_000)
    y: int = Field(ge=-2_048, le=2_048)
    z: int = Field(ge=-30_000_000, le=30_000_000)


class JoinArguments(SafeModel):
    connection_profile_id: ProfileId | None = None


class LeaveArguments(SafeModel):
    reason: SafeReason | None = None


class FollowOwnerArguments(SafeModel):
    follow_distance: float = Field(default=3.0, ge=2.0, le=6.0)
    lease_duration_seconds: int = Field(default=300, ge=5, le=900)


class WaitHereArguments(SafeModel):
    reason: SafeReason | None = None


class ComeHereArguments(SafeModel):
    arrival_distance: float = Field(default=3.0, ge=2.0, le=4.0)


class LookAtOwnerArguments(SafeModel):
    duration_seconds: float = Field(default=2.0, ge=0.1, le=10.0)


class ReportStatusArguments(SafeModel):
    detail_level: Literal["basic", "standard"] = "standard"


class StopArguments(SafeModel):
    reason: SafeReason | None = None


class EmergencyStopArguments(SafeModel):
    reason: SafeReason | None = None


class CommandPayloadBase(SafeModel):
    deadline_at: AwareDatetime

    @field_validator("deadline_at")
    @classmethod
    def deadline_must_be_utc(cls, value: datetime) -> datetime:
        return _require_utc(value, "deadline_at")


class JoinCommandPayload(CommandPayloadBase):
    name: Literal["join"]
    arguments: JoinArguments = Field(default_factory=JoinArguments)


class LeaveCommandPayload(CommandPayloadBase):
    name: Literal["leave"]
    arguments: LeaveArguments = Field(default_factory=LeaveArguments)


class FollowOwnerCommandPayload(CommandPayloadBase):
    name: Literal["follow_owner"]
    arguments: FollowOwnerArguments = Field(default_factory=FollowOwnerArguments)


class WaitHereCommandPayload(CommandPayloadBase):
    name: Literal["wait_here"]
    arguments: WaitHereArguments = Field(default_factory=WaitHereArguments)


class ComeHereCommandPayload(CommandPayloadBase):
    name: Literal["come_here"]
    arguments: ComeHereArguments = Field(default_factory=ComeHereArguments)


class LookAtOwnerCommandPayload(CommandPayloadBase):
    name: Literal["look_at_owner"]
    arguments: LookAtOwnerArguments = Field(default_factory=LookAtOwnerArguments)


class ReportStatusCommandPayload(CommandPayloadBase):
    name: Literal["report_status"]
    arguments: ReportStatusArguments = Field(default_factory=ReportStatusArguments)


class StopCommandPayload(CommandPayloadBase):
    name: Literal["stop"]
    arguments: StopArguments = Field(default_factory=StopArguments)


class EmergencyStopCommandPayload(CommandPayloadBase):
    name: Literal["emergency_stop"]
    arguments: EmergencyStopArguments = Field(default_factory=EmergencyStopArguments)


CommandPayload = Annotated[
    JoinCommandPayload
    | LeaveCommandPayload
    | FollowOwnerCommandPayload
    | WaitHereCommandPayload
    | ComeHereCommandPayload
    | LookAtOwnerCommandPayload
    | ReportStatusCommandPayload
    | StopCommandPayload
    | EmergencyStopCommandPayload,
    Field(discriminator="name"),
]


class ProtocolEnvelope(SafeModel):
    protocol: Literal["gamma.minecraft"]
    version: Literal[1]
    type: str
    message_id: BoundedId
    sent_at: AwareDatetime
    sequence: int = Field(ge=0)

    @field_validator("sent_at")
    @classmethod
    def sent_at_must_be_utc(cls, value: datetime) -> datetime:
        return _require_utc(value, "sent_at")

    @field_validator("version", "sequence", mode="before")
    @classmethod
    def integer_fields_must_not_be_coerced(cls, value: object) -> object:
        if type(value) is not int:
            raise ValueError("version and sequence must be JSON integers")
        return value


class ConnectedEnvelope(ProtocolEnvelope):
    connection_id: BoundedId


class CommandEnvelope(ConnectedEnvelope):
    trace_id: BoundedId
    command_id: BoundedId


class HelloPayload(SafeModel):
    supported_versions: list[Literal[1]] = Field(min_length=1, max_length=4)
    sidecar_instance_id: BoundedId
    sidecar_build: BoundedVersion
    capabilities: list[Literal["companion_v1"]] = Field(min_length=1, max_length=4)
    node_version: BoundedVersion
    minecraft_library_version: BoundedVersion
    pathfinder_version: BoundedVersion
    companion_state: Literal[CompanionState.DISCONNECTED, CompanionState.IDLE]


class WelcomePayload(SafeModel):
    selected_version: Literal[1]
    heartbeat_interval_seconds: int = Field(ge=1, le=60)
    liveness_timeout_seconds: int = Field(ge=2, le=180)
    maximum_message_bytes: int = Field(ge=1_024, le=1_048_576)
    command_cache_ttl_seconds: int = Field(ge=60, le=86_400)
    command_cache_capacity: int = Field(ge=1, le=10_000)
    minecraft_chat_output_enabled: Literal[False] = False


class HeartbeatPayload(SafeModel):
    companion_state: CompanionState
    last_received_sequence: int | None = Field(default=None, ge=0)


class SidecarStatusPayload(SafeModel):
    connection_state: SidecarConnectionState
    companion_state: CompanionState
    uptime_seconds: int = Field(ge=0, le=31_536_000)
    last_failure: Failure | None = None


class MinecraftStatusPayload(SafeModel):
    connection_state: MinecraftConnectionState
    companion_state: CompanionState
    negotiated_version: BoundedVersion | None = None
    dimension: MinecraftDimension | None = None


class CommandAckPayload(SafeModel):
    accepted: bool
    command_name: CommandName
    failure: Failure | None = None

    @model_validator(mode="after")
    def failure_matches_acceptance(self) -> Self:
        if self.accepted and self.failure is not None:
            raise ValueError("accepted acknowledgments cannot contain a failure")
        if not self.accepted and self.failure is None:
            raise ValueError("rejected acknowledgments require a failure")
        return self


class CommandProgressPayload(SafeModel):
    command_name: CommandName
    phase: Literal["started", "moving", "waiting", "retrying"]
    elapsed_ms: int = Field(ge=0, le=86_400_000)
    safe_detail: SafeDetail | None = None


class StateSnapshotPayload(SafeModel):
    sidecar_connection_state: SidecarConnectionState
    minecraft_connection_state: MinecraftConnectionState
    companion_state: CompanionState
    owner_present: bool
    owner_display_name: SafeDisplayName | None = None
    owner_uuid: UUID | None = None
    dimension: MinecraftDimension | None = None
    rounded_position: RoundedPosition | None = None
    health: float | None = Field(default=None, ge=0.0, le=20.0)
    hunger: int | None = Field(default=None, ge=0, le=20)
    active_command_id: BoundedId | None = None
    active_command_name: CommandName | None = None
    last_terminal_outcome: TerminalOutcome | None = None
    last_failure_code: FailureCode | None = None

    @model_validator(mode="after")
    def active_command_fields_are_paired(self) -> Self:
        if (self.active_command_id is None) != (self.active_command_name is None):
            raise ValueError("active command identifier and name must appear together")
        return self


class TerminalResultPayload(SafeModel):
    command_name: CommandName
    outcome: TerminalOutcome
    elapsed_ms: int = Field(ge=0, le=86_400_000)
    safe_detail: SafeDetail | None = None
    failure: Failure | None = None

    @model_validator(mode="after")
    def failure_matches_outcome(self) -> Self:
        failure_outcomes = {
            TerminalOutcome.FAILED,
            TerminalOutcome.REJECTED,
            TerminalOutcome.TIMED_OUT,
        }
        if self.outcome in failure_outcomes and self.failure is None:
            raise ValueError("failed, rejected, and timed-out results require a failure")
        if self.outcome not in failure_outcomes and self.failure is not None:
            raise ValueError("completed and cancelled results cannot contain a failure")
        return self


class CancelCommandPayload(SafeModel):
    reason: SafeReason | None = None


class EmergencyStopPayload(SafeModel):
    reason: SafeReason | None = None


class ShutdownPayload(SafeModel):
    reason: SafeReason | None = None
    leave_minecraft: Literal[True] = True


class ProtocolErrorPayload(SafeModel):
    code: FailureCode
    safe_detail: SafeDetail
    retriable: bool
    offending_type: Annotated[str, StringConstraints(min_length=1, max_length=64)] | None = None


class HelloMessage(ProtocolEnvelope):
    type: Literal["hello"]
    payload: HelloPayload


class WelcomeMessage(ConnectedEnvelope):
    type: Literal["welcome"]
    payload: WelcomePayload


class HeartbeatMessage(ConnectedEnvelope):
    type: Literal["heartbeat"]
    payload: HeartbeatPayload


class SidecarStatusMessage(ConnectedEnvelope):
    type: Literal["sidecar_status"]
    payload: SidecarStatusPayload


class MinecraftStatusMessage(ConnectedEnvelope):
    type: Literal["minecraft_status"]
    payload: MinecraftStatusPayload


class CommandMessage(CommandEnvelope):
    type: Literal["command"]
    payload: CommandPayload

    @model_validator(mode="after")
    def deadline_follows_send_time(self) -> Self:
        if self.payload.deadline_at <= self.sent_at:
            raise ValueError("deadline_at must be later than sent_at")
        if self.payload.deadline_at - self.sent_at > timedelta(
            seconds=MAX_COMMAND_DEADLINE_SECONDS
        ):
            raise ValueError("deadline_at cannot be more than 900 seconds after sent_at")
        return self


class CommandAckMessage(CommandEnvelope):
    type: Literal["command_ack"]
    payload: CommandAckPayload


class CommandProgressMessage(CommandEnvelope):
    type: Literal["command_progress"]
    payload: CommandProgressPayload


class StateSnapshotMessage(ConnectedEnvelope):
    type: Literal["state_snapshot"]
    trace_id: BoundedId | None = None
    command_id: BoundedId | None = None
    payload: StateSnapshotPayload

    @model_validator(mode="after")
    def command_correlation_is_complete(self) -> Self:
        if self.command_id is not None and self.trace_id is None:
            raise ValueError("command-correlated snapshots require trace_id")
        return self


class TerminalResultMessage(CommandEnvelope):
    type: Literal["terminal_result"]
    payload: TerminalResultPayload


class CancelCommandMessage(CommandEnvelope):
    type: Literal["cancel_command"]
    payload: CancelCommandPayload


class EmergencyStopMessage(CommandEnvelope):
    type: Literal["emergency_stop"]
    payload: EmergencyStopPayload


class ShutdownMessage(ConnectedEnvelope):
    type: Literal["shutdown"]
    trace_id: BoundedId
    payload: ShutdownPayload


class ProtocolErrorMessage(ProtocolEnvelope):
    type: Literal["protocol_error"]
    connection_id: BoundedId | None = None
    trace_id: BoundedId | None = None
    command_id: BoundedId | None = None
    payload: ProtocolErrorPayload


ProtocolMessage = Annotated[
    HelloMessage
    | WelcomeMessage
    | HeartbeatMessage
    | SidecarStatusMessage
    | MinecraftStatusMessage
    | CommandMessage
    | CommandAckMessage
    | CommandProgressMessage
    | StateSnapshotMessage
    | TerminalResultMessage
    | CancelCommandMessage
    | EmergencyStopMessage
    | ShutdownMessage
    | ProtocolErrorMessage,
    Field(discriminator="type"),
]

PROTOCOL_MESSAGE_ADAPTER = TypeAdapter(ProtocolMessage)


def parse_protocol_message(data: str | bytes | bytearray) -> ProtocolMessage:
    """Parse one JSON protocol message through the v1 discriminator."""

    return PROTOCOL_MESSAGE_ADAPTER.validate_json(data)


def _require_utc(value: datetime, field_name: str) -> datetime:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError(f"{field_name} must be timezone-aware")
    if value.utcoffset() != timedelta(0):
        raise ValueError(f"{field_name} must use UTC")
    return value
